### PR TITLE
Availability/Extrinsic store

### LINKED ISSUE #371: Availability of parachain data
(without economic games)

## Part 1: Data store
We would want to plug this into the client's RocksDB instance somehow, and answer `BlockData` requests out of it.

In the consensus worker, we shouldn't cast an `Available` vote until we've actually ensured availability of the data. 

https://github.com/paritytech/polkadot/blob/ddc8d705570fda5f492895373b69b9b84cae1d69/polkadot/consensus/src/shared_table/mod.rs#L256-L257

should be preceded by a call to `store.make_available(block_data, extrinsic)` or something similar.

For now, just storing block data, indexed by `(Hash, ParaId)`, the relay chain block hash and parachain ID.

## Part 2: Network protocol for fetching data

This would be used as a last resort in case the data doesn't appear to be easily available.
Finding validator who is bonded to have the data and requesting from them. The validator answers out of the data store

We would add this in the `PolkadotProtocol`, probably with a trait object `DataStore`

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #371 

Introduces an availability store. Validators who are meant to guarantee available data will not cast votes until the data has been written to disk. Currently it resides in a fixed subdirectory of the general data directory, but we will make it configurable in the future.

Block data network requests are now answered from the store if not kept in memory already.
There is no protocol for discovering peer IDs for validators who have historic data available (yet).

### CODE REVIEW

File: polkadot/availability-store/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+version = "0.1.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+
+[dependencies]
+polkadot-primitives = { path = "../primitives" }
+parking_lot = "0.4"
+log = "0.3"
+substrate-codec = { path = "../../substrate/codec" }
+substrate-primitives = { path = "../../substrate/primitives" }
+kvdb = { git = "https://github.com/paritytech/parity.git" }
```
Comment (dvdplm) [inline polkadot/availability-store/Cargo.toml]: The `kvdb*` crates are in `parity-common` now.

File: polkadot/consensus/src/shared_table/mod.rs [hunk 2]
```diff
@@ -251,12 +264,11 @@ impl<D, E, C, Err> Future for PrimedStatementProducer<D, E, C>
...
-				let hash = work.candidate_receipt.hash();
+				let hash = candidate_hash();
 				debug!(target: "consensus", "Claiming candidate {} available.", hash);
 
-				// TODO: actually wait for block data and then ensure availability.
-				self.inner.produced_statements.extrinsic = Some(extrinsic);
-				self.inner.produced_statements.availability =
+				statements.extrinsic = Some(extrinsic);
+				statements.availability =
 					Some(GenericStatement::Available(hash));
```
Comment (gnunicorn) [inline polkadot/consensus/src/shared_table/mod.rs]: wrap unnecessary now.

File: polkadot/consensus/src/service.rs [hunk 3]
```diff
@@ -84,10 +135,11 @@ impl Service {
 		thread_pool: ThreadPoolHandle,
 		parachain_empty_duration: Duration,
 		key: ed25519::Pair,
+		extrinsic_store: ExtrinsicStore,
 	) -> Service
 		where
 			A: LocalPolkadotApi + Send + Sync + 'static,
-			C: BlockchainEvents<Block> + ChainHead<Block> + bft::BlockImport<Block> + bft::Authorities<Block> + Send + Sync + 'static,
+			C: BlockchainEvents<Block> + ChainHead<Block> + BlockBody<Block> + bft::BlockImport<Block> + bft::Authorities<Block> + Send + Sync + 'static,
```
Comment (gnunicorn) [inline polkadot/consensus/src/service.rs]: can we wrap this?

File: polkadot/availability-store/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	/// Block data.
+	pub block_data: BlockData,
+	/// Extrinsic data.
+	pub extrinsic: Option<Extrinsic>,
+}
+
+fn extract_io_err(err: ::kvdb::Error) -> io::Error {
+	match err {
+		::kvdb::Error(::kvdb::ErrorKind::Io(io_err), _) => io_err,
+		::kvdb::Error(::kvdb::ErrorKind::Msg(msg), _) => io::Error::new(
```
Comment (dvdplm) [inline polkadot/availability-store/src/lib.rs]: These errors are gone and replaced (in `parity-ethereum`) by plain `io::Error`. [PR](https://github.com/paritytech/parity-ethereum/pull/8924)

File: polkadot/availability-store/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+pub struct Store {
+	inner: Arc<KeyValueDB>,
+}
+
+impl Store {
+	/// Create a new `Store` with given config on disk.
+	pub fn new(config: Config) -> io::Result<Self> {
+		let mut db_config = DatabaseConfig::with_columns(Some(columns::NUM_COLUMNS));
+		db_config.memory_budget = config.cache_size;
+		db_config.wal = true;
```
Comment (dvdplm) [inline polkadot/availability-store/src/lib.rs]: This option is also gone (always true now).

File: polkadot/availability-store/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+	/// Create a new `Store` in-memory. Useful for tests.
+	pub fn new_in_memory() -> Self {
+		Store {
+			inner: Arc::new(::kvdb_memorydb::create(::columns::NUM_COLUMNS)),
+		}
+	}
+
+	/// Make some data available provisionally.
+	pub fn make_available(&self, data: Data) -> io::Result<()> {
```
Comment (dvdplm) [inline polkadot/availability-store/src/lib.rs]: What's wrong with `fetch` or `get`?

**Pull request merged**